[libc++] Granularize <vector>

philnik777 opened the pull request:
|                    | old time | new time |
| ------------------ | -------- | -------- |
| functional - c++23 | 416ms    | 225ms    |
| random - c++23     | 513ms    | 392ms    |
| vector - c++17     | 206ms    | 100ms    |

philnik777 committed:
[libc++] Granularize <vector>

cjdb reviewed (changes requested):
What is the motivation for this patch?

There were two goals when I originally started the modularisation of libc++ headers:

1. eliminate circular dependencies
2. separate completely unrelated entities

This was done with the intention of ensuring that including Header A didn't include all of Header B just for Function F, and to improve lookup (`<type_traits>` was the home of many things that it wasn't the prescribed home of).

Splitting `<vector>` into seven new headers doesn't achieve the first goal, and I'm not convinced it is achieving the second either.

ldionne commented:
> What is the motivation for this patch?

It allows reducing the compilation times for `<vector>` which spiked after we introduced the formatters.

When we introduced formatters, we got many reports of people finding their compilation times and preprocessed file sizes regressed by a huge amount. That was tracked down to the fact that `<vector>` was now including a bunch of stuff for `<format>`. By splitting up `<vector>` into subparts like this, other parts of the library can include only the parts that they actually need.

FWIW, I think we could potentially get away with a less granular split. For example, we could only extract the formatters in one header and have everything else in another header, both of which would be included from `<vector>`. If we want to split `vector` and `vector<bool>`, we have to have comparison operators, `swap` and `remove` in a shared location too. We could maybe do `__vector/common.h`?

cjdb commented:
> > What is the motivation for this patch?
> 
> It allows reducing the compilation times for `<vector>` which spiked after we introduced the formatters.
> 
> When we introduced formatters, we got many reports of people finding their compilation times and preprocessed file sizes regressed by a huge amount. That was tracked down to the fact that `<vector>` was now including a bunch of stuff for `<format>`. By splitting up `<vector>` into subparts like this, other parts of the library can include only the parts that they actually need.

Thanks for the context. I would appreciate it if the commit message incorporated this information, as it isn't currently clear why this change is important.

> FWIW, I think we could potentially get away with a less granular split. For example, we could only extract the formatters in one header and have everything else in another header, both of which would be included from `<vector>`. If we want to split `vector` and `vector<bool>`, we have to have comparison operators, `swap` and `remove` in a shared location too. We could maybe do `__vector/common.h`?

I'd appreciate seeing the results of an A/B experiment before splitting out `vector<bool>` into its own file (as well as generally understanding how we're benchmarking this). I'd also like to see this information into the commit message, or mentioning an associated GH issue in the commit message that documents the results and processed.

philnik777 commented:
> > > What is the motivation for this patch?
> > 
> > 
> > It allows reducing the compilation times for `<vector>` which spiked after we introduced the formatters.
> > When we introduced formatters, we got many reports of people finding their compilation times and preprocessed file sizes regressed by a huge amount. That was tracked down to the fact that `<vector>` was now including a bunch of stuff for `<format>`. By splitting up `<vector>` into subparts like this, other parts of the library can include only the parts that they actually need.
> 
> Thanks for the context. I would appreciate it if the commit message incorporated this information, as it isn't currently clear why this change is important.
> 
> > FWIW, I think we could potentially get away with a less granular split. For example, we could only extract the formatters in one header and have everything else in another header, both of which would be included from `<vector>`. If we want to split `vector` and `vector<bool>`, we have to have comparison operators, `swap` and `remove` in a shared location too. We could maybe do `__vector/common.h`?
> 
> I'd appreciate seeing the results of an A/B experiment before splitting out `vector<bool>` into its own file (as well as generally understanding how we're benchmarking this). I'd also like to see this information into the commit message, or mentioning an associated GH issue in the commit message that documents the results and processed.

Splitting out `vector<bool>` reduces the include times of `<functional>` by ~18ms. I'm getting the numbers through `-ftime-trace`.

ldionne commented on libcxx/include/__vector/container_traits.h:
```diff
@@ -0,0 +1,39 @@
+_LIBCPP_BEGIN_NAMESPACE_STD
+
+template <class _Tp, class _Allocator>
+struct __container_traits<vector<_Tp, _Allocator> > {
```
I am a bit concerned about putting this here separately. Right now, I don't think this gets included anywhere. So if we want to check for `__container_traits<_SomeContainer>` in a generic context (such as from `flat_map`, which is where this trait originated from), we'd currently never get a specialization for `std::vector` AFAICT.

It's kind of weird, but in a certain way I'd imagine that `__vector/vector.h` should perhaps do something like this:

```c++
namespace std {
  // define std::vector
}

#include <__vector/swap.h>
#include <__vector/comparison.h>
#include <__vector/container_traits.h>
// etc.
```

And `<__vector/vector_bool.h>` would do similarly:

```
namespace std {
  // define std::vector<bool>
}

#include <__vector/swap.h>
#include <__vector/comparison.h>
#include <__vector/container_traits.h>
// etc.
```

That way, you'd be certain to have all the associated operations when you include `<__vector/vector.h>`. I think we have the same issue for `swap`, except it might be more subtle. If you have the wrong includes, you might end up calling the generic `std::swap` which is probably going to compile but is also less efficient.

ldionne replied:
I didn't look properly. You have includes of these headers at the top of `vector.h` and `vector_bool.h`.

ldionne commented on libcxx/include/__vector/formatter.h:
```diff
@@ -0,0 +1,49 @@
+#if _LIBCPP_STD_VER >= 23
+
+_LIBCPP_BEGIN_NAMESPACE_STD
+
```
Should the `__enable_insertable` specialization live here instead?

ldionne replied:
Answer: no, since that would require including this header from `vector.h`, and avoiding this inclusion is the whole point of this patch.

ldionne commented on libcxx/include/__vector/vector.h:
```diff
@@ -0,0 +1,1534 @@
+#include <__utility/move.h>
+#include <__utility/pair.h>
+#include <__utility/swap.h>
+#include <__vector/comparison.h>
+#include <__vector/container_traits.h>
+#include <__vector/swap.h>
```
We could make these includes stand out more with a small comment saying they are part of the definition.

ldionne commented on libcxx/include/__vector/vector_bool.h:
```diff
@@ -0,0 +1,1121 @@
+#include <__utility/forward.h>
+#include <__utility/move.h>
+#include <__utility/swap.h>
+#include <__vector/comparison.h>
+#include <__vector/container_traits.h>
+#include <__vector/swap.h>
```
Same here.

ldionne commented on libcxx/include/__vector/formatter.h:
```diff
@@ -0,0 +1,49 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _LIBCPP___VECTOR_FORMATTER_H
```
We should rename this to `vector_bool_formatter.h` since it's only the formatter for `vector<bool>`.

ldionne reviewed (approved):
LGTM with my non-resolved comments.

ldionne reviewed (approved):
LGTM with green CI.
